In [103]:
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np

In [104]:
df = pd.read_csv("owid-covid-data.csv")
df

iso_code continent     location        date  total_cases  new_cases  \
0           AFG      Asia  Afghanistan  2020-02-24          5.0        5.0   
1           AFG      Asia  Afghanistan  2020-02-25          5.0        0.0   
2           AFG      Asia  Afghanistan  2020-02-26          5.0        0.0   
3           AFG      Asia  Afghanistan  2020-02-27          5.0        0.0   
4           AFG      Asia  Afghanistan  2020-02-28          5.0        0.0   
...         ...       ...          ...         ...          ...        ...   
166321      ZWE    Africa     Zimbabwe  2022-03-01     236871.0      491.0   
166322      ZWE    Africa     Zimbabwe  2022-03-02     237503.0      632.0   
166323      ZWE    Africa     Zimbabwe  2022-03-03     237503.0        0.0   
166324      ZWE    Africa     Zimbabwe  2022-03-04     238739.0     1236.0   
166325      ZWE    Africa     Zimbabwe  2022-03-05     239019.0      280.0   

        new_cases_smoothed  total_deaths  new_deaths  new_deaths_smoothed  \
0                      NaN           NaN         NaN                  NaN   
1                      NaN           NaN         NaN                  NaN   
2                      NaN           NaN         NaN                  NaN   
3                      NaN           NaN         NaN                  NaN   
4                      NaN           NaN         NaN                  NaN   
...                    ...           ...         ...                  ...   
166321             413.000        5395.0         0.0                1.000   
166322             416.286        5396.0         1.0                1.143   
166323             362.286        5396.0         0.0                0.857   
166324             467.429        5397.0         1.0                0.714   
166325             459.429        5397.0         0.0                0.571   

        ...  female_smokers  male_smokers  handwashing_facilities  \
0       ...             NaN           NaN                  37.746   
1       ...             NaN           NaN                  37.746   
2       ...             NaN           NaN                  37.746   
3       ...             NaN           NaN                  37.746   
4       ...             NaN           NaN                  37.746   
...     ...             ...           ...                     ...   
166321  ...             1.6          30.7                  36.791   
166322  ...             1.6          30.7                  36.791   
166323  ...             1.6          30.7                  36.791   
166324  ...             1.6          30.7                  36.791   
166325  ...             1.6          30.7                  36.791   

        hospital_beds_per_thousand  life_expectancy  human_development_index  \
0                              0.5            64.83                    0.511   
1                              0.5            64.83                    0.511   
2                              0.5            64.83                    0.511   
3                              0.5            64.83                    0.511   
4                              0.5            64.83                    0.511   
...                            ...              ...                      ...   
166321                         1.7            61.49                    0.571   
166322                         1.7            61.49                    0.571   
166323                         1.7            61.49                    0.571   
166324                         1.7            61.49                    0.571   
166325                         1.7            61.49                    0.571   

        excess_mortality_cumulative_absolute  excess_mortality_cumulative  \
0                                        NaN                          NaN   
1                                        NaN                          NaN   
2                                        NaN                          NaN   
3                                        NaN   

In [105]:
USA_data = df[df['location'] == 'United States']
USA_data = USA_data.set_index('date')
Brazil_data = df[df['location'] == 'Brazil']
Brazil_data = Brazil_data.set_index('date')
Germany_data = df[df['location'] == 'Germany']
Germany_data = Germany_data.set_index('date')
China_data = df[df['location'] == 'China']
China_data = China_data.set_index('date')
Nigeria_data = df[df['location'] == 'Nigeria']
Nigeria_data = Nigeria_data.set_index('date')
Australia_data = df[df['location'] == 'Australia']
Australia_data = Australia_data.set_index('date')

In [106]:
USA_train = USA_data[0:710]
USA_test = USA_data[711:800]
Brazil_train = Brazil_data[0:710]
Brazil_test = Brazil_data[711:800]
Germany_train = Germany_data[0:710]
Germany_test = Germany_data[711:800]
China_train = China_data[0:710]
China_test = China_data[711:800]
Nigeria_train = Nigeria_data[0:710]
Nigeria_test = Nigeria_data[711:800]
Australia_train = Australia_data[0:710]
Australia_test = Australia_data[711:800]

In [107]:
df.columns

Index(['iso_code', 'continent', 'location', 'date', 'total_cases', 'new_cases',
       'new_cases_smoothed', 'total_deaths', 'new_deaths',
       'new_deaths_smoothed', 'total_cases_per_million',
       'new_cases_per_million', 'new_cases_smoothed_per_million',
       'total_deaths_per_million', 'new_deaths_per_million',
       'new_deaths_smoothed_per_million', 'reproduction_rate', 'icu_patients',
       'icu_patients_per_million', 'hosp_patients',
       'hosp_patients_per_million', 'weekly_icu_admissions',
       'weekly_icu_admissions_per_million', 'weekly_hosp_admissions',
       'weekly_hosp_admissions_per_million', 'new_tests', 'total_tests',
       'total_tests_per_thousand', 'new_tests_per_thousand',
       'new_tests_smoothed', 'new_tests_smoothed_per_thousand',
       'positive_rate', 'tests_per_case', 'tests_units', 'total_vaccinations',
       'people_vaccinated', 'people_fully_vaccinated', 'total_boosters',
       'new_vaccinations', 'new_vaccinations_smoothed',
       't

# Vaccinations

Many countries are missing vaccination data only within an initial time period (about a 1 year delay), but are generally complete going forward. Each of the countries will be check to verify this information, and may be chosen as predictors despite higher missingness proportions. We anticipate our models to be robust to this specific pattern of missingness.

# USA

In [108]:
missing_USA = pd.DataFrame(USA_train.isnull().sum())
missing_USA['% missing'] = USA_train.isnull().sum()/USA_train.shape[0]
missing_USA.columns = ["Missing Count", "Proportion Missing"]
missing_USA = missing_USA.sort_values(by="Proportion Missing", ascending=True)
missing_USA = missing_USA[missing_USA["Proportion Missing"] < .50]
print(missing_USA.shape)
missing_USA

(57, 2)


Missing Count  Proportion Missing
iso_code                                                0            0.000000
population_density                                      0            0.000000
median_age                                              0            0.000000
aged_65_older                                           0            0.000000
aged_70_older                                           0            0.000000
gdp_per_capita                                          0            0.000000
extreme_poverty                                         0            0.000000
cardiovasc_death_rate                                   0            0.000000
diabetes_prevalence                                     0            0.000000
female_smokers                                          0            0.000000
male_smokers                                            0            0.000000
hospital_beds_per_thousand                              0            0.000000
life_expectancy                                         0            0.000000
stringency_index                                        0            0.000000
total_cases_per_million                                 0            0.000000
human_development_index                                 0            0.000000
total_cases                                             0            0.000000
location                                                0            0.000000
continent                                               0            0.000000
population                                              0            0.000000
new_cases_per_million                                   1            0.001408
new_cases                                               1            0.001408
new_cases_smoothed_per_million                          7            0.009859
new_cases_smoothed                                      7            0.009859
total_deaths_per_million                               38            0.053521
new_deaths                                             38            0.053521
total_deaths                                           38            0.053521
new_deaths_per_million                                 38            0.053521
tests_units                                            39            0.054930
total_tests_per_thousand                               39            0.054930
new_tests                                              39            0.054930
total_tests                                            39            0.054930
new_tests_per_thousand                                 39            0.054930
reproduction_rate                                      43            0.060563
new_deaths_smoothed                                    44            0.061972
new_deaths_smoothed_per_million                        44            0.061972
positive_rate                                          45            0.063380
tests_per_case                                         45            0.063380
new_tests_smoothed                                     46            0.064789
new_tests_smoothed_per_thousand                        46            0.064789
icu_patients                                          175            0.246479
icu_patients_per_million                              175            0.246479
hosp_patients                                         175            0.246479
hosp_patients_per_million                             175            0.246479
weekly_hosp_admissions_per_million                    181            0.254930
weekly_hosp_admissions                                181            0.254930
total_vaccinations_per_hundred                        326            0.459155
people_vaccinated                                     326            0.459155
people_fully_vaccinated                               326            0.459155
people_vaccinated_per_hundred                         326            0.459155
people_fully_vaccinated_per_hundred                   326         

In [109]:
# Selecting predictors with adequate missingness

selected_USA = missing_USA.index
USA_train = USA_train[selected_USA] 
USA_test = USA_test[selected_USA] 

# Brazil

In [110]:
missing_Brazil = pd.DataFrame(Brazil_train.isnull().sum())
missing_Brazil['% missing'] = Brazil_train.isnull().sum()/Brazil_train.shape[0]
missing_Brazil.columns = ["Missing Count", "Proportion Missing"]
missing_Brazil = missing_Brazil.sort_values(by="Proportion Missing", ascending=True)
missing_Brazil = missing_Brazil[missing_Brazil["Proportion Missing"] < .50]
print(missing_Brazil.shape)
missing_Brazil

(47, 2)


Missing Count  Proportion Missing
iso_code                                                0            0.000000
population_density                                      0            0.000000
median_age                                              0            0.000000
aged_65_older                                           0            0.000000
aged_70_older                                           0            0.000000
gdp_per_capita                                          0            0.000000
extreme_poverty                                         0            0.000000
cardiovasc_death_rate                                   0            0.000000
diabetes_prevalence                                     0            0.000000
female_smokers                                          0            0.000000
male_smokers                                            0            0.000000
hospital_beds_per_thousand                              0            0.000000
life_expectancy                                         0            0.000000
human_development_index                                 0            0.000000
population                                              0            0.000000
stringency_index                                        0            0.000000
total_cases_per_million                                 0            0.000000
total_cases                                             0            0.000000
location                                                0            0.000000
continent                                               0            0.000000
new_cases_per_million                                   2            0.002817
new_cases                                               2            0.002817
reproduction_rate                                      17            0.023944
new_deaths_per_million                                 20            0.028169
total_deaths_per_million                               20            0.028169
new_cases_smoothed_per_million                         20            0.028169
new_deaths                                             20            0.028169
total_deaths                                           20            0.028169
new_cases_smoothed                                     20            0.028169
new_deaths_smoothed                                    26            0.036620
new_deaths_smoothed_per_million                        26            0.036620
tests_units                                            60            0.084507
new_tests_smoothed                                     67            0.094366
new_tests_smoothed_per_thousand                        67            0.094366
new_vaccinations_smoothed                             327            0.460563
new_vaccinations_smoothed_per_million                 327            0.460563
new_people_vaccinated_smoothed                        327            0.460563
new_people_vaccinated_smoothed_per_hundred            327            0.460563
total_vaccinations                                    332            0.467606
total_vaccinations_per_hundred                        332            0.467606
new_vaccinations                                      335            0.471831
people_vaccinated                                     336            0.473239
people_vaccinated_per_hundred                         336            0.473239
people_fully_vaccinated                               352            0.495775
people_fully_vaccinated_per_hundred                   352            0.495775
total_tests_per_thousand                              353            0.497183
total_tests                                           353            0.497183

In [111]:
# Selecting predictors with adequate missingness

selected_Brazil = missing_Brazil.index[:-2]
Brazil_train = Brazil_train[selected_Brazil] 
Brazil_test = Brazil_test[selected_Brazil] 

# Germany

In [112]:
missing_Germany = pd.DataFrame(Germany_train.isnull().sum())
missing_Germany['% missing'] = Germany_train.isnull().sum()/Germany_train.shape[0]
missing_Germany.columns = ["Missing Count", "Proportion Missing"]
missing_Germany = missing_Germany.sort_values(by="Proportion Missing", ascending=True)
missing_Germany = missing_Germany[missing_Germany["Proportion Missing"] < .50]
print(missing_Germany.shape)
missing_Germany

(50, 2)


Missing Count  Proportion Missing
iso_code                                                0            0.000000
population_density                                      0            0.000000
median_age                                              0            0.000000
aged_65_older                                           0            0.000000
aged_70_older                                           0            0.000000
gdp_per_capita                                          0            0.000000
cardiovasc_death_rate                                   0            0.000000
diabetes_prevalence                                     0            0.000000
female_smokers                                          0            0.000000
male_smokers                                            0            0.000000
hospital_beds_per_thousand                              0            0.000000
population                                              0            0.000000
life_expectancy                                         0            0.000000
stringency_index                                        0            0.000000
continent                                               0            0.000000
location                                                0            0.000000
total_cases_per_million                                 0            0.000000
human_development_index                                 0            0.000000
total_cases                                             0            0.000000
new_cases_per_million                                   2            0.002817
new_cases                                               2            0.002817
new_cases_smoothed                                     20            0.028169
new_cases_smoothed_per_million                         20            0.028169
weekly_hosp_admissions                                 34            0.047887
weekly_hosp_admissions_per_million                     34            0.047887
reproduction_rate                                      35            0.049296
total_deaths                                           42            0.059155
total_deaths_per_million                               42            0.059155
new_deaths                                             46            0.064789
new_deaths_per_million                                 46            0.064789
tests_units                                            48            0.067606
icu_patients_per_million                               53            0.074648
icu_patients                                           53            0.074648
new_tests_smoothed                                     55            0.077465
new_tests_smoothed_per_thousand                        55            0.077465
new_deaths_smoothed_per_million                        76            0.107042
new_deaths_smoothed                                    76            0.107042
total_vaccinations                                    335            0.471831
people_vaccinated                                     335            0.471831
people_fully_vaccinated                               335            0.471831
total_vaccinations_per_hundred                        335            0.471831
people_vaccinated_per_hundred                         335            0.471831
people_fully_vaccinated_per_hundred                   335            0.471831
new_people_vaccinated_smoothed_per_hundred            336            0.473239
new_vaccinations_smoothed_per_million                 336            0.473239
new_vaccinations_smoothed                             336            0.473239
new_vaccinations                                      336            0.473239
new_people_vaccinated_smoothed                        336            0.473239
total_boosters_per_hundred                            337            0.474648
total_boosters                                        337            0.474648

In [113]:
# Selecting predictors with adequate missingness

selected_Germany = missing_Germany.index
Germany_train = Germany_train[selected_Germany] 
Germany_test = Germany_test[selected_Germany] 

# China

In [114]:
missing_China = pd.DataFrame(China_train.isnull().sum())
missing_China['% missing'] = China_train.isnull().sum()/China_train.shape[0]
missing_China.columns = ["Missing Count", "Proportion Missing"]
missing_China = missing_China.sort_values(by="Proportion Missing", ascending=True)
missing_China = missing_China[missing_China["Proportion Missing"] < .70]
print(missing_China.shape)
missing_China

(36, 2)


Missing Count  Proportion Missing
iso_code                                           0            0.000000
stringency_index                                   0            0.000000
population                                         0            0.000000
population_density                                 0            0.000000
median_age                                         0            0.000000
aged_65_older                                      0            0.000000
aged_70_older                                      0            0.000000
gdp_per_capita                                     0            0.000000
extreme_poverty                                    0            0.000000
cardiovasc_death_rate                              0            0.000000
diabetes_prevalence                                0            0.000000
female_smokers                                     0            0.000000
male_smokers                                       0            0.000000
life_expectancy                                    0            0.000000
human_development_index                            0            0.000000
hospital_beds_per_thousand                         0            0.000000
continent                                          0            0.000000
total_deaths_per_million                           0            0.000000
location                                           0            0.000000
total_cases_per_million                            0            0.000000
total_deaths                                       0            0.000000
total_cases                                        0            0.000000
reproduction_rate                                  1            0.001408
new_deaths_per_million                             2            0.002817
new_deaths                                         2            0.002817
new_cases_per_million                              3            0.004225
new_cases                                          3            0.004225
new_deaths_smoothed_per_million                   14            0.019718
new_deaths_smoothed                               14            0.019718
new_cases_smoothed                                21            0.029577
new_cases_smoothed_per_million                    21            0.029577
new_vaccinations_smoothed                        329            0.463380
new_vaccinations_smoothed_per_million            329            0.463380
total_vaccinations                               413            0.581690
total_vaccinations_per_hundred                   413            0.581690
new_vaccinations                                 426            0.600000

In [115]:
# Selecting predictors with adequate missingness


selected_China = missing_China.index
China_train = China_train[selected_China] 
China_test = China_test[selected_China] 

# Nigeria

In [116]:
missing_Nigeria = pd.DataFrame(Nigeria_train.isnull().sum())
missing_Nigeria['% missing'] = Nigeria_train.isnull().sum()/Nigeria_train.shape[0]
missing_Nigeria.columns = ["Missing Count", "Proportion Missing"]
missing_Nigeria = missing_Nigeria.sort_values(by="Proportion Missing", ascending=True)
missing_Nigeria = missing_Nigeria[missing_Nigeria["Proportion Missing"] < .60]
print(missing_Nigeria.shape)
missing_Nigeria

(39, 2)


Missing Count  Proportion Missing
iso_code                                                0            0.000000
population_density                                      0            0.000000
median_age                                              0            0.000000
aged_65_older                                           0            0.000000
aged_70_older                                           0            0.000000
gdp_per_capita                                          0            0.000000
cardiovasc_death_rate                                   0            0.000000
diabetes_prevalence                                     0            0.000000
female_smokers                                          0            0.000000
male_smokers                                            0            0.000000
handwashing_facilities                                  0            0.000000
life_expectancy                                         0            0.000000
human_development_index                                 0            0.000000
population                                              0            0.000000
stringency_index                                        0            0.000000
new_cases_per_million                                   0            0.000000
total_cases_per_million                                 0            0.000000
new_cases                                               0            0.000000
total_cases                                             0            0.000000
location                                                0            0.000000
continent                                               0            0.000000
new_cases_smoothed_per_million                          6            0.008451
new_cases_smoothed                                      6            0.008451
total_deaths_per_million                               24            0.033803
total_deaths                                           24            0.033803
new_deaths                                             25            0.035211
new_deaths_per_million                                 25            0.035211
reproduction_rate                                      31            0.043662
new_deaths_smoothed                                    37            0.052113
new_deaths_smoothed_per_million                        37            0.052113
tests_units                                            40            0.056338
tests_per_case                                         47            0.066197
positive_rate                                          47            0.066197
new_tests_smoothed                                     47            0.066197
new_tests_smoothed_per_thousand                        47            0.066197
new_vaccinations_smoothed                             371            0.522535
new_people_vaccinated_smoothed_per_hundred            371            0.522535
new_people_vaccinated_smoothed                        371            0.522535
new_vaccinations_smoothed_per_million                 371            0.522535

In [117]:
# Selecting predictors with adequate missingness

selected_Nigeria = missing_Nigeria.index
Nigeria_train = Nigeria_train[selected_Nigeria] 
Nigeria_test = Nigeria_test[selected_Nigeria] 

# Australia

In [118]:
missing_Australia = pd.DataFrame(Australia_train.isnull().sum())
missing_Australia['% missing'] = Australia_train.isnull().sum()/Australia_train.shape[0]
missing_Australia.columns = ["Missing Count", "Proportion Missing"]
missing_Australia = missing_Australia.sort_values(by="Proportion Missing", ascending=True)
missing_Australia = missing_Australia[missing_Australia["Proportion Missing"] < .60]
print(missing_Australia.shape)
missing_Australia

(55, 2)


Missing Count  Proportion Missing
iso_code                                                0            0.000000
population_density                                      0            0.000000
median_age                                              0            0.000000
aged_65_older                                           0            0.000000
aged_70_older                                           0            0.000000
gdp_per_capita                                          0            0.000000
extreme_poverty                                         0            0.000000
cardiovasc_death_rate                                   0            0.000000
diabetes_prevalence                                     0            0.000000
female_smokers                                          0            0.000000
male_smokers                                            0            0.000000
hospital_beds_per_thousand                              0            0.000000
life_expectancy                                         0            0.000000
stringency_index                                        0            0.000000
new_cases_per_million                                   0            0.000000
total_cases_per_million                                 0            0.000000
human_development_index                                 0            0.000000
new_cases                                               0            0.000000
total_cases                                             0            0.000000
location                                                0            0.000000
continent                                               0            0.000000
population                                              0            0.000000
new_cases_smoothed_per_million                          6            0.008451
new_cases_smoothed                                      6            0.008451
total_deaths_per_million                               35            0.049296
total_deaths                                           35            0.049296
new_deaths                                             38            0.053521
new_deaths_per_million                                 38            0.053521
tests_units                                            41            0.057746
total_tests_per_thousand                               41            0.057746
total_tests                                            41            0.057746
new_tests                                              42            0.059155
new_tests_per_thousand                                 42            0.059155
reproduction_rate                                      45            0.063380
new_tests_smoothed                                     48            0.067606
new_tests_smoothed_per_thousand                        48            0.067606
positive_rate                                          48            0.067606
tests_per_case                                         48            0.067606
new_deaths_smoothed_per_million                        62            0.087324
new_deaths_smoothed                                    62            0.087324
icu_patients_per_million                               65            0.091549
hosp_patients                                          65            0.091549
hosp_patients_per_million                              65            0.091549
icu_patients                                           65            0.091549
new_vaccinations_smoothed                             393            0.553521
new_vaccinations_smoothed_per_million                 393            0.553521
new_people_vaccinated_smoothed                        393            0.553521
new_people_vaccinated_smoothed_per_hundred            393            0.553521
total_vaccinations                                    397            0.559155
people_vaccinated                                     397            0.559155
people_vaccinated_per_hundred                         397         

In [119]:
# Selecting predictors with adequate missingness

selected_Australia = missing_Australia.index
Australia_train = Australia_train[selected_Australia] 
Australia_test = Australia_test[selected_Australia] 

# Notes, Selection Justification

* Though we are checking predictor missingness off of the training data set, we are looking at the actual data spread from the entire CSV file in excel. This way, we can check the date ranges the missingness falls within, if there is a pattern.

* Final predictor count for each country is 1 less than listed from the missingness data frames generated, because we are exclusing iso_code as a predictor.


1. USA_train has 45 predictors below 30% missingness. The next 11 predictors had missingness between 45-46%. However, these predictors displayed missingness only up until middle of December 2020, so they are kept. Thus, 56 predictors are kept for USA, demonstrating up to 46% missingness. There is a large gap between the additional set of predictors and the original set, whose highest missingness was around 25%. (56)

2. Brazil_train only has 33 predictors below 30% missingness. The next 13 predictors had missingness between 46-49%. However, 11 of these predictors displayed missingness only up until early February of 2021, so they are kept. The other two predictors (total_tests and total_tests_per_thousand) had sporadic missingness all across the timeline, so they were removed. Thus, 44 predictors are kept for Brazil, demonstrating upto 47.3% missingness. There is a large gap between the additional set of predictors and the original set, whose highest missingness was around 9.4%. (44)

3. Germany_train has only 36 predictors below 30% missingness. The next 13 predictors had missingness around 47%. However, these predictors only displayed missingess up until the end of December 2020, so they are kept. Thus, 49 predictors are kept for Germany, demonstrating missingness up to 47%. There is a large gap between the additional set of predictors and the original set, whose highest missingness was around 10%. (49)

4. China_train only has 30 predictors below 30% missingness. The next 2 are new_vaccinations_smoothed and new_vaccinations_smoothed_per_million, which had missingness around 46%. However, these two only displayed missingness up until the middle of December 2020, so they are kept. Though the the other vaccination based predictors had missingness about 50%, they are kept because they stay below 60%. Thus, a total of 35 predictors are selected. (35)

5. Nigeria_train has only 34 predictors below 30% missingness. Though the next 4 predictors following these display missingness above 50%, so they fall below 60%, so they are kept. Thus, we will select a total of 38 predictors. (38)

6. Australia_train has 43 predictors below 30% missingness. Even though the following 11 predictors are above 50% missingness, they are below 60%, so threy are kept. Thus, we will select a total of 54 predictors. (54)

# Addressing Missingness and Imputing values

After evaluating the definitions of all the variables, our group decided to drop the three following variables despite their proportion of missing values being 0%: diabetes_prevalance, female_smokers, male_smokers. The reasoning behind is that none of three variables have much of an intuitive relationship or correlation with the response variable new_cases. Additionally, all the values for a column are the same in every row, which will coontribute less meaningful data into our analysis and eventual model-buildings process. Therefore, we have dropped these three columns from all six countries.

In [120]:
USA_train = USA_train.drop(columns = ['diabetes_prevalence', 'female_smokers', 'male_smokers'], axis=1)
USA_test = USA_test.drop(columns = ['diabetes_prevalence', 'female_smokers', 'male_smokers'], axis=1)

Brazil_train = Brazil_train.drop(columns = ['diabetes_prevalence', 'female_smokers', 'male_smokers'], axis=1)
Brazil_test = Brazil_test.drop(columns = ['diabetes_prevalence', 'female_smokers', 'male_smokers'], axis=1)

Germany_train = Germany_train.drop(columns = ['diabetes_prevalence', 'female_smokers', 'male_smokers'], axis=1)
Germany_test = Germany_test.drop(columns = ['diabetes_prevalence', 'female_smokers', 'male_smokers'], axis=1)

China_train = China_train.drop(columns = ['diabetes_prevalence', 'female_smokers', 'male_smokers'], axis=1)
China_test = China_test.drop(columns = ['diabetes_prevalence', 'female_smokers', 'male_smokers'], axis=1)

Nigeria_train = Nigeria_train.drop(columns = ['diabetes_prevalence', 'female_smokers', 'male_smokers'], axis=1)
Nigeria_test = Nigeria_test.drop(columns = ['diabetes_prevalence', 'female_smokers', 'male_smokers'], axis=1)

Australia_train = Australia_train.drop(columns = ['diabetes_prevalence', 'female_smokers', 'male_smokers'], axis=1)
Australia_test = Australia_test.drop(columns = ['diabetes_prevalence', 'female_smokers', 'male_smokers'], axis=1)

# After dropping by contextual intuition and reasoning

1. USA_train now has 42 predictors below 30% missingness. The next 11 predictors had missingness between 45-46%. However, these predictors displayed missingness only up until middle of December 2020, so they are kept. Thus, 56 predictors are kept for USA, demonstrating up to 46% missingness. There is a large gap between the additional set of predictors and the original set, whose highest missingness was around 25%. (53)

2. Brazil_train now has 30 predictors below 30% missingness. The next 13 predictors had missingness between 46-49%. However, 11 of these predictors displayed missingness only up until early February of 2021, so they are kept. The other two predictors (total_tests and total_tests_per_thousand) had sporadic missingness all across the timeline, so they were removed. Thus, 44 predictors are kept for Brazil, demonstrating upto 47.3% missingness. There is a large gap between the additional set of predictors and the original set, whose highest missingness was around 9.4%. (41)

3. Germany_train now has 33 predictors below 30% missingness. The next 13 predictors had missingness around 47%. However, these predictors only displayed missingess up until the end of December 2020, so they are kept. Thus, 49 predictors are kept for Germany, demonstrating missingness up to 47%. There is a large gap between the additional set of predictors and the original set, whose highest missingness was around 10%. (46)

4. China_train now has 27 predictors below 30% missingness. The next 2 are new_vaccinations_smoothed and new_vaccinations_smoothed_per_million, which had missingness around 46%. However, these two only displayed missingness up until the middle of December 2020, so they are kept. Though the the other vaccination based predictors had missingness about 50%, they are kept because they stay below 60%. Thus, a total of 35 predictors are selected. (32)

5. Nigeria_train now has 31 predictors below 30% missingness. Though the next 4 predictors following these display missingness above 50%, so they fall below 60%, so they are kept. Thus, we will select a total of 35 predictors. (35)

6. Australia_train now has 40 predictors below 30% missingness. Even though the following 11 predictors are above 50% missingness, they are below 60%, so threy are kept. Thus, we will select a total of 51 predictors. (51)

# Imputing missing values
There are two approaches we will follow when imputing for missing data. The first is manually setting the missing values for any vaccination, death or case-related variables to 0 for all entries in the early dates of 2020 from the train data. The reasoning behind this is because many of the variables with missing values for these entries simply were not created until later in 2020. For example, new_cases and new_cases_smoothed could only be attained as testing for Covid became more accessible and a widespread practice. Setting these values to 0 is most appropritate for the misisng values in 2020. Any other misisng values for these variables were filled using forward fill. The second approach involves filling the missing values for any other predictor variables not previously described using backward fill. The reasoning and justification for this is that reproduction_rate and icu_pateints might not have been known at the time, but these values were not 0. Therefore, using backward will is most appropriate because we fill all the values of early 2020 with the first recorded entry. After 2020, we also used forward fill to impute any remaining data. In the event where data was missing throughout/randomly over time but not from the very beginning of a country's dataset, we used ffill to impute the missing values.

## United States

In [121]:
# variables that have missing values for first part of 2020 -- filling missing values in 2020 with 0
columns_to_fill_0 = ['new_cases_per_million', 'new_cases', 'new_cases_smoothed_per_million', 'new_cases_smoothed', 
                     'total_deaths_per_million', 'new_deaths', 'total_deaths', 'new_deaths_per_million', 'tests_units', 
                     'total_tests_per_thousand', 'new_tests', 'total_tests', 'new_tests_per_thousand', 
                     'new_deaths_smoothed', 'new_deaths_smoothed_per_million', 'positive_rate', 'tests_per_case', 
                     'new_tests_smoothed', 'new_tests_smoothed_per_thousand', 'total_vaccinations_per_hundred', 
                     'people_vaccinated', 'people_fully_vaccinated', 'people_vaccinated_per_hundred', 'people_fully_vaccinated_per_hundred', 
                     'total_vaccinations', 'new_vaccinations', 'new_vaccinations_smoothed', 'new_people_vaccinated_smoothed_per_hundred', 
                     'new_people_vaccinated_smoothed', 'new_vaccinations_smoothed_per_million']

# variables that have missing values for first part of 2020 -- filling missing values in 2020 with forward bfill
columns_to_bfill = ['reproduction_rate', 'icu_patients', 'icu_patients_per_million', 'hosp_patients', 
                     'hosp_patients_per_million', 'weekly_hosp_admissions_per_million', 'weekly_hosp_admissions']             

# Applying zero-fill for the first 340 rows (all the rows of 2020 train data) for columns_to_fill_0
USA_train[columns_to_fill_0] = USA_train[columns_to_fill_0].fillna(0, limit=340)

# Applying forward fill for remaining rows (dates after 2020) for columns_to_fill_0
USA_train[columns_to_fill_0] = USA_train[columns_to_fill_0].fillna(method='ffill')

# Applying backward fill for the first 340 rows (all the rows of 2020 train data) for columns_to_bfill
USA_train[columns_to_bfill] = USA_train[columns_to_bfill].fillna(method='bfill', limit=340)

# Applying forward fill for the remaining rows (dates after 2020) for columns_to_bfill
USA_train[columns_to_bfill] = USA_train[columns_to_bfill].fillna(method='ffill')

In [122]:
# Check if there are any missing values in the entire DataFrame
any_missing = USA_train.isna().any().any()  # or df.isnull().any().any()

# Print the result
print("Are there any missing values in the United States DataFrame?", any_missing)

Are there any missing values in the United States DataFrame? False


## Brazil

In [123]:
# variables that have missing values for first part of 2020 -- filling missing values in 2020 with 0
columns_to_fill_0 = ['new_deaths_per_million', 'total_deaths_per_million', 'new_cases_smoothed_per_million', 'new_deaths', 
                     'total_deaths', 'new_cases_smoothed', 'new_deaths_smoothed', 'new_deaths_smoothed_per_million', 
                     'tests_units', 'new_tests_smoothed', 'new_tests_smoothed_per_thousand', 'new_vaccinations_smoothed',
                     'new_vaccinations_smoothed_per_million', 'new_people_vaccinated_smoothed', 'new_people_vaccinated_smoothed_per_hundred',
                     'total_vaccinations', 'total_vaccinations_per_hundred', 'new_vaccinations', 'people_vaccinated', 
                     'people_vaccinated_per_hundred', 'people_fully_vaccinated', 'people_fully_vaccinated_per_hundred']

# variables that have missing values for first part of 2020 -- filling missing values in 2020 with forward bfill
columns_to_bfill = ['reproduction_rate']
    
# variables that have missing values for first part of 2020 -- filling missing values in 2020 with forward bfill
columns_to_ffill = ['new_cases_per_million', 'new_cases']

# Applying zero-fill for the first 100 rows (all the rows of early 2020 train with missing data) for columns_to_fill_0
Brazil_train[columns_to_fill_0] = Brazil_train[columns_to_fill_0].fillna(0, limit=100)

# Applying forward fill for remaining rows (dates after 2020) for columns_to_fill_0
Brazil_train[columns_to_fill_0] = Brazil_train[columns_to_fill_0].fillna(method='ffill')

# Applying backward fill for the first 100 rows (all the rows of early 2020 train with missing data) for columns_to_bfill
Brazil_train[columns_to_bfill] = Brazil_train[columns_to_bfill].fillna(method='bfill', limit=100)

# Applying forward fill for the remaining rows (dates after 2020) for columns_to_bfill
Brazil_train[columns_to_bfill] = Brazil_train[columns_to_bfill].fillna(method='ffill')

# Applying forward fill for new_cases_per_million and new_cases
Brazil_train[columns_to_ffill] = Brazil_train[columns_to_ffill].fillna(method='ffill')

In [124]:
# Check if there are any missing values in the entire DataFrame
any_missing = Brazil_train.isna().any().any()  # or df.isnull().any().any()

# Print the result
print("Are there any missing values in the Brazil DataFrame?", any_missing)

Are there any missing values in the Brazil DataFrame? False


## Germany

In [125]:
# variables that have missing values for first part of 2020 -- filling missing values in 2020 with 0
columns_to_fill_0 = ['total_vaccinations', 'people_vaccinated', 'people_fully_vaccinated', 'total_vaccinations_per_hundred',
                    'people_vaccinated_per_hundred', 'people_fully_vaccinated_per_hundred', 'new_people_vaccinated_smoothed_per_hundred',
                    'new_vaccinations_smoothed_per_million', 'new_vaccinations_smoothed', 'new_vaccinations', 'new_people_vaccinated_smoothed',
                    'total_boosters_per_hundred', 'total_boosters', 'new_cases_smoothed', 'new_cases_smoothed_per_million', 
                    'weekly_hosp_admissions', 'weekly_hosp_admissions_per_million', 'total_deaths', 'total_deaths_per_million', 
                    'new_deaths', 'new_deaths_per_million', 'tests_units', 'new_tests_smoothed', 
                    'new_tests_smoothed_per_thousand', 'new_deaths_smoothed_per_million', 'new_deaths_smoothed']

# variables that have missing values for first part of 2020 -- filling missing values in 2020 with forward bfill
columns_to_bfill = ['reproduction_rate', 'icu_patients', 'icu_patients_per_million',
                    'weekly_hosp_admissions_per_million', 'weekly_hosp_admissions']             

# variables that have missing values for first part of 2020 -- filling missing values in 2020 with forward bfill
columns_to_ffill = ['new_cases_per_million', 'new_cases']


# Applying zero-fill for the first 100 rows (all the rows of early 2020 train with missing data) for columns_to_fill_0
Germany_train[columns_to_fill_0] = Germany_train[columns_to_fill_0].fillna(0, limit=100)

# Applying forward fill for remaining rows (dates after 2020) for columns_to_fill_0
Germany_train[columns_to_fill_0] = Germany_train[columns_to_fill_0].fillna(method='ffill')

# Applying backward fill for the first 100 rows (all the rows of early 2020 train with missing data) for columns_to_bfill
Germany_train[columns_to_bfill] = Germany_train[columns_to_bfill].fillna(method='bfill', limit=100)

# Applying forward fill for the remaining rows (dates after 2020) for columns_to_bfill
Germany_train[columns_to_bfill] = Germany_train[columns_to_bfill].fillna(method='ffill')

# Applying forward fill for new_cases_per_million and new_cases
Germany_train[columns_to_ffill] = Germany_train[columns_to_ffill].fillna(method='ffill')

In [126]:
# Check if there are any missing values in the entire DataFrame
any_missing = Germany_train.isna().any().any()  # or df.isnull().any().any()

# Print the result
print("Are there any missing values in the Germany DataFrame?", any_missing)

Are there any missing values in the Germany DataFrame? False


## China

In [127]:
# variables that have missing values for first part of 2020 (until June) -- filling missing values in 2020 with 0 
columns_to_fill_0_p1 = ['new_deaths_per_million', 'new_deaths', 'new_cases_per_million', 'new_cases', 
                     'new_deaths_smoothed_per_million', 'new_deaths_smoothed', 'new_cases_smoothed',
                     'new_cases_smoothed_per_million']

# variables that have missing values for most part of 2020 -- filling missing values in 2020 with 0 
columns_to_fill_0_p2 = ['new_vaccinations_smoothed', 'new_vaccinations_smoothed_per_million', 'new_vaccinations']

# variables that have missing values for most part of 2020 -- filling missing values in 2020 with 0
columns_to_fill_0_p3 = ['total_vaccinations']

# variables that have missing values for most part of 2020 -- filling missing values in 2020 with 0 
columns_to_fill_0_p4 = ['total_vaccinations_per_hundred']

# variables that have missing values for first part of 2020 -- filling missing values in 2020 with forward bfill
columns_to_bfill = ['reproduction_rate']             

# Applying zero-fill for the first 100 rows (all the rows of early 2020 train with missing data) for columns_to_fill_0_p1
China_train[columns_to_fill_0_p1] = China_train[columns_to_fill_0_p1].fillna(0, limit=100)

# Applying forward fill for remaining rows (dates after 2020) for columns_to_fill_0_p1
China_train[columns_to_fill_0_p1] = China_train[columns_to_fill_0_p1].fillna(method='ffill')

# Applying zero-fill for the first 440 rows (all the rows of early 2020 train with missing data) for columns_to_fill_0_p2
China_train[columns_to_fill_0_p2] = China_train[columns_to_fill_0_p2].fillna(0, limit=440)

# Applying forward fill for remaining rows (dates after 2020) for columns_to_fill_0_p1
China_train[columns_to_fill_0_p2] = China_train[columns_to_fill_0_p2].fillna(method='ffill')

# Applying zero-fill for the first 329 rows (all the rows of early 2020 train with missing data) for columns_to_fill_0_p3
China_train[columns_to_fill_0_p3] = China_train[columns_to_fill_0_p3].fillna(0, limit=329)

# Applying forward fill for remaining rows (dates after 2020) for columns_to_fill_0_p1
China_train[columns_to_fill_0_p3] = China_train[columns_to_fill_0_p3].fillna(method='ffill')

# Applying zero-fill for the first 330 rows (all the rows of early 2020 train with missing data) for columns_to_fill_0_p4
China_train[columns_to_fill_0_p4] = China_train[columns_to_fill_0_p4].fillna(0, limit=330)

# Applying forward fill for remaining rows (dates after 2020) for columns_to_fill_0_p1
China_train[columns_to_fill_0_p4] = China_train[columns_to_fill_0_p4].fillna(method='ffill')

# Applying backward fill for the first 100 rows (all the rows of early 2020 train with missing data) for columns_to_bfill
China_train[columns_to_bfill] = China_train[columns_to_bfill].fillna(method='bfill')

In [128]:
# Check if there are any missing values in the entire DataFrame
any_missing = China_train.isna().any().any()  # or df.isnull().any().any()

# Print the result
print("Are there any missing values in the China DataFrame?", any_missing)

Are there any missing values in the China DataFrame? False


## Nigeria

In [129]:
# variables that have missing values for first part of 2020 -- filling missing values in 2020 with 0
columns_to_fill_0 = ['new_cases_smoothed_per_million', 'new_cases_smoothed', 'total_deaths_per_million', 
                     'total_deaths', 'new_deaths', 'new_deaths_per_million', 'new_deaths_smoothed',
                     'new_deaths_smoothed_per_million', 'tests_units', 'tests_per_case', 'positive_rate', 
                     'new_tests_smoothed', 'new_tests_smoothed_per_thousand', 'new_vaccinations_smoothed', 
                     'new_people_vaccinated_smoothed_per_hundred', 'new_people_vaccinated_smoothed', 
                     'new_vaccinations_smoothed_per_million'] 

# variables that have missing values for first part of 2020 -- filling missing values in 2020 with forward bfill
columns_to_bfill = ['reproduction_rate']             

# Applying zero-fill for the first 340 rows (all the rows of 2020 train data) for columns_to_fill_0
Nigeria_train[columns_to_fill_0] = Nigeria_train[columns_to_fill_0].fillna(0, limit=380)

# Applying forward fill for remaining rows (dates after 2020) for columns_to_fill_0
Nigeria_train[columns_to_fill_0] = Nigeria_train[columns_to_fill_0].fillna(method='ffill')

# Applying backward fill for the first 340 rows (all the rows of 2020 train data) for columns_to_bfill
Nigeria_train[columns_to_bfill] = Nigeria_train[columns_to_bfill].fillna(method='bfill')

In [130]:
# Check if there are any missing values in the entire DataFrame
any_missing = Nigeria_train.isna().any().any()  # or df.isnull().any().any()

# Print the result
print("Are there any missing values in the Nigeria DataFrame?", any_missing)

Are there any missing values in the Nigeria DataFrame? False


## Australia

In [131]:
# variables that have missing values for first part of 2020 -- filling missing values in 2020 with 0
columns_to_fill_0_p1 = ['new_cases_smoothed_per_million', 'new_cases_smoothed', 'total_deaths_per_million', 'total_deaths',
                        'new_deaths', 'new_deaths_per_million', 'tests_units', 'total_tests_per_thousand', 'total_tests',
                        'new_tests', 'new_tests_per_thousand', 'new_tests_smoothed', 'new_tests_smoothed_per_thousand', 
                        'positive_rate', 'tests_per_case', 'new_deaths_smoothed_per_million', 'new_deaths_smoothed']                 

# variables that have missing values for first part of 2020 -- filling missing values in 2020 with 0
columns_to_fill_0_p2 = ['new_vaccinations_smoothed', 'new_vaccinations_smoothed_per_million', 'new_people_vaccinated_smoothed', 
                        'new_people_vaccinated_smoothed_per_hundred', 'total_vaccinations', 'people_vaccinated', 
                        'people_vaccinated_per_hundred', 'total_vaccinations_per_hundred', 'people_fully_vaccinated',
                        'people_fully_vaccinated_per_hundred', 'new_vaccinations']

# variables that have missing values for first part of 2020 -- filling missing values in 2020 with forward bfill
columns_to_bfill = ['reproduction_rate', 'icu_patients', 'icu_patients_per_million', 'hosp_patients', 
                     'hosp_patients_per_million']             

# Applying zero-fill for the first 50 rows (all the rows of 2020 train data) for columns_to_fill_0
Australia_train[columns_to_fill_0_p1] = Australia_train[columns_to_fill_0_p1].fillna(0, limit=50)

# Applying forward fill for remaining rows (dates after 2020) for columns_to_fill_0
Australia_train[columns_to_fill_0_p1] = Australia_train[columns_to_fill_0_p1].fillna(method='ffill')

# Applying zero-fill for the first 400 rows (all the rows of 2020 train data) for columns_to_fill_0
Australia_train[columns_to_fill_0_p2] = Australia_train[columns_to_fill_0_p2].fillna(0, limit=400)

# Applying forward fill for remaining rows (dates after 2020) for columns_to_fill_0
Australia_train[columns_to_fill_0_p2] = Australia_train[columns_to_fill_0_p2].fillna(method='ffill')

# Applying backward fill for the first 70 rows (all the rows of 2020 train data) for columns_to_bfill
Australia_train[columns_to_bfill] = Australia_train[columns_to_bfill].fillna(method='bfill', limit=70)

# Applying forward fill for the remaining rows (dates after 2020) for columns_to_bfill
Australia_train[columns_to_bfill] = Australia_train[columns_to_bfill].fillna(method='ffill')

In [132]:
# Check if there are any missing values in the entire DataFrame
any_missing = Australia_train.isna().any().any()  # or df.isnull().any().any()

# Print the result
print("Are there any missing values in the Australia DataFrame?", any_missing)

Are there any missing values in the Australia DataFrame? False


In [133]:
USA_train

iso_code  population_density  median_age  aged_65_older  \
date                                                                 
2020-01-22      USA              35.608        38.3         15.413   
2020-01-23      USA              35.608        38.3         15.413   
2020-01-24      USA              35.608        38.3         15.413   
2020-01-25      USA              35.608        38.3         15.413   
2020-01-26      USA              35.608        38.3         15.413   
...             ...                 ...         ...            ...   
2021-12-27      USA              35.608        38.3         15.413   
2021-12-28      USA              35.608        38.3         15.413   
2021-12-29      USA              35.608        38.3         15.413   
2021-12-30      USA              35.608        38.3         15.413   
2021-12-31      USA              35.608        38.3         15.413   

            aged_70_older  gdp_per_capita  extreme_poverty  \
date                                                         
2020-01-22          9.732       54225.446              1.2   
2020-01-23          9.732       54225.446              1.2   
2020-01-24          9.732       54225.446              1.2   
2020-01-25          9.732       54225.446              1.2   
2020-01-26          9.732       54225.446              1.2   
...                   ...             ...              ...   
2021-12-27          9.732       54225.446              1.2   
2021-12-28          9.732       54225.446              1.2   
2021-12-29          9.732       54225.446              1.2   
2021-12-30          9.732       54225.446              1.2   
2021-12-31          9.732       54225.446              1.2   

            cardiovasc_death_rate  hospital_beds_per_thousand  \
date                                                            
2020-01-22                151.089                        2.77   
2020-01-23                151.089                        2.77   
2020-01-24                151.089                        2.77   
2020-01-25                151.089                        2.77   
2020-01-26                151.089                        2.77   
...                           ...                         ...   
2021-12-27                151.089                        2.77   
2021-12-28                151.089                        2.77   
2021-12-29                151.089                        2.77   
2021-12-30                151.089                        2.77   
2021-12-31                151.089                        2.77   

            life_expectancy  ...  people_vaccinated  people_fully_vaccinated  \
date                         ...                                               
2020-01-22            78.86  ...                0.0                      0.0   
2020-01-23            78.86  ...                0.0                      0.0   
2020-01-24            78.86  ...                0.0                      0.0   
2020-01-25            78.86  ...                0.0                      0.0   
2020-01-26            78.86  ...                0.0                      0.0   
...                     ...  ...                ...                      ...   
2021-12-27            78.86  ...        242268651.0              206735879.0   
2021-12-28            78.86  ...        242702682.0              206973527.0   
2021-12-29            78.86  ...        243137810.0              207206943.0   
2021-12-30            78.86  ...        243540028.0              207418330.0   
2021-12-31            78.86  ...        243701693.0              207510289.0   

            people_vaccinated_per_hundred  \
date                                        
2020-01-22                           0.00   
2020-01-23                           0.00   
2020-01-24                           0.00   
2020-01-25                           0.00   
2020-01-26                           0.00   
...                                   ...   
2021-12-27                          72.97   
2

In [134]:
Brazil_train

iso_code  population_density  median_age  aged_65_older  \
date                                                                 
2020-02-26      BRA               25.04        33.5          8.552   
2020-02-27      BRA               25.04        33.5          8.552   
2020-02-28      BRA               25.04        33.5          8.552   
2020-02-29      BRA               25.04        33.5          8.552   
2020-03-01      BRA               25.04        33.5          8.552   
...             ...                 ...         ...            ...   
2022-01-31      BRA               25.04        33.5          8.552   
2022-02-01      BRA               25.04        33.5          8.552   
2022-02-02      BRA               25.04        33.5          8.552   
2022-02-03      BRA               25.04        33.5          8.552   
2022-02-04      BRA               25.04        33.5          8.552   

            aged_70_older  gdp_per_capita  extreme_poverty  \
date                                                         
2020-02-26           5.06       14103.452              3.4   
2020-02-27           5.06       14103.452              3.4   
2020-02-28           5.06       14103.452              3.4   
2020-02-29           5.06       14103.452              3.4   
2020-03-01           5.06       14103.452              3.4   
...                   ...             ...              ...   
2022-01-31           5.06       14103.452              3.4   
2022-02-01           5.06       14103.452              3.4   
2022-02-02           5.06       14103.452              3.4   
2022-02-03           5.06       14103.452              3.4   
2022-02-04           5.06       14103.452              3.4   

            cardiovasc_death_rate  hospital_beds_per_thousand  \
date                                                            
2020-02-26                177.961                         2.2   
2020-02-27                177.961                         2.2   
2020-02-28                177.961                         2.2   
2020-02-29                177.961                         2.2   
2020-03-01                177.961                         2.2   
...                           ...                         ...   
2022-01-31                177.961                         2.2   
2022-02-01                177.961                         2.2   
2022-02-02                177.961                         2.2   
2022-02-03                177.961                         2.2   
2022-02-04                177.961                         2.2   

            life_expectancy  ...  new_vaccinations_smoothed_per_million  \
date                         ...                                          
2020-02-26            75.88  ...                                    0.0   
2020-02-27            75.88  ...                                    0.0   
2020-02-28            75.88  ...                                    0.0   
2020-02-29            75.88  ...                                    0.0   
2020-03-01            75.88  ...                                    0.0   
...                     ...  ...                                    ...   
2022-01-31            75.88  ...                                 5090.0   
2022-02-01            75.88  ...                                 5234.0   
2022-02-02            75.88  ...                                 5328.0   
2022-02-03            75.88  ...                                 5139.0   
2022-02-04            75.88  ...                                 4864.0   

            new_people_vaccinated_smoothed  \
date                                         
2020-02-26                             0.0   
2020-02-27                             0.0   
2020-02-28                             0.0   
2020-02-29                             0.0   
2020-03-01                             0.0   
...                                    ...   
2022-01-31                        283001.0   
2022-02-01                        293247.0   
2022-02-02 

In [135]:
Germany_train

iso_code  population_density  median_age  aged_65_older  \
date                                                                 
2020-01-27      DEU             237.016        46.6         21.453   
2020-01-28      DEU             237.016        46.6         21.453   
2020-01-29      DEU             237.016        46.6         21.453   
2020-01-30      DEU             237.016        46.6         21.453   
2020-01-31      DEU             237.016        46.6         21.453   
...             ...                 ...         ...            ...   
2022-01-01      DEU             237.016        46.6         21.453   
2022-01-02      DEU             237.016        46.6         21.453   
2022-01-03      DEU             237.016        46.6         21.453   
2022-01-04      DEU             237.016        46.6         21.453   
2022-01-05      DEU             237.016        46.6         21.453   

            aged_70_older  gdp_per_capita  cardiovasc_death_rate  \
date                                                               
2020-01-27         15.957       45229.245                156.139   
2020-01-28         15.957       45229.245                156.139   
2020-01-29         15.957       45229.245                156.139   
2020-01-30         15.957       45229.245                156.139   
2020-01-31         15.957       45229.245                156.139   
...                   ...             ...                    ...   
2022-01-01         15.957       45229.245                156.139   
2022-01-02         15.957       45229.245                156.139   
2022-01-03         15.957       45229.245                156.139   
2022-01-04         15.957       45229.245                156.139   
2022-01-05         15.957       45229.245                156.139   

            hospital_beds_per_thousand  population  life_expectancy  ...  \
date                                                                 ...   
2020-01-27                         8.0  83900471.0            81.33  ...   
2020-01-28                         8.0  83900471.0            81.33  ...   
2020-01-29                         8.0  83900471.0            81.33  ...   
2020-01-30                         8.0  83900471.0            81.33  ...   
2020-01-31                         8.0  83900471.0            81.33  ...   
...                                ...         ...              ...  ...   
2022-01-01                         8.0  83900471.0            81.33  ...   
2022-01-02                         8.0  83900471.0            81.33  ...   
2022-01-03                         8.0  83900471.0            81.33  ...   
2022-01-04                         8.0  83900471.0            81.33  ...   
2022-01-05                         8.0  83900471.0            81.33  ...   

            total_vaccinations_per_hundred people_vaccinated_per_hundred  \
date                                                                       
2020-01-27                            0.00                          0.00   
2020-01-28                            0.00                          0.00   
2020-01-29                            0.00                          0.00   
2020-01-30                            0.00                          0.00   
2020-01-31                            0.00                          0.00   
...                                    ...                           ...   
2022-01-01                          179.12                         73.69   
2022-01-02                          179.30                         73.71   
2022-01-03                          179.87                         73.76   
2022-01-04                          180.71                         73.83   
2022-01-05                          181.68                         73.90   

           people_fully_vaccinated_per_hundred  \
date                                             
2020-01-27                                0.00   
2020-01-28                                0.00   
2020-01-29                           

In [136]:
China_train

iso_code  stringency_index    population  population_density  \
date                                                                      
2020-01-22      CHN             26.39  1.444216e+09             147.674   
2020-01-23      CHN             44.91  1.444216e+09             147.674   
2020-01-24      CHN             44.91  1.444216e+09             147.674   
2020-01-25      CHN             47.69  1.444216e+09             147.674   
2020-01-26      CHN             69.91  1.444216e+09             147.674   
...             ...               ...           ...                 ...   
2021-12-27      CHN             79.17  1.444216e+09             147.674   
2021-12-28      CHN             79.17  1.444216e+09             147.674   
2021-12-29      CHN             79.17  1.444216e+09             147.674   
2021-12-30      CHN             79.17  1.444216e+09             147.674   
2021-12-31      CHN             79.17  1.444216e+09             147.674   

            median_age  aged_65_older  aged_70_older  gdp_per_capita  \
date                                                                   
2020-01-22        38.7         10.641          5.929       15308.712   
2020-01-23        38.7         10.641          5.929       15308.712   
2020-01-24        38.7         10.641          5.929       15308.712   
2020-01-25        38.7         10.641          5.929       15308.712   
2020-01-26        38.7         10.641          5.929       15308.712   
...                ...            ...            ...             ...   
2021-12-27        38.7         10.641          5.929       15308.712   
2021-12-28        38.7         10.641          5.929       15308.712   
2021-12-29        38.7         10.641          5.929       15308.712   
2021-12-30        38.7         10.641          5.929       15308.712   
2021-12-31        38.7         10.641          5.929       15308.712   

            extreme_poverty  cardiovasc_death_rate  ...  new_cases  \
date                                                ...              
2020-01-22              0.7                261.899  ...        0.0   
2020-01-23              0.7                261.899  ...       92.0   
2020-01-24              0.7                261.899  ...      277.0   
2020-01-25              0.7                261.899  ...      483.0   
2020-01-26              0.7                261.899  ...      663.0   
...                     ...                    ...  ...        ...   
2021-12-27              0.7                261.899  ...      361.0   
2021-12-28              0.7                261.899  ...      197.0   
2021-12-29              0.7                261.899  ...      207.0   
2021-12-30              0.7                261.899  ...      195.0   
2021-12-31              0.7                261.899  ...      229.0   

            new_deaths_smoothed_per_million  new_deaths_smoothed  \
date                                                               
2020-01-22                              0.0                  0.0   
2020-01-23                              0.0                  0.0   
2020-01-24                              0.0                  0.0   
2020-01-25                              0.0                  0.0   
2020-01-26                              0.0                  0.0   
...                                     ...                  ...   
2021-12-27                              0.0                  0.0   
2021-12-28                              0.0                  0.0   
2021-12-29                              0.0                  0.0   
2021-12-30                              0.0                  0.0   
2021-12-31                              0.0                  0.0   

           new_cases_smoothed  new_cases_smoothed_per_million  \
date                                                            
2020-01-22              0.000                           0.000   
2020-01-23              0.000                           0.000   
2020-01-24              0.000             

In [137]:
Nigeria_train

iso_code  population_density  median_age  aged_65_older  \
date                                                                 
2020-02-28      NGA             209.588        18.1          2.751   
2020-02-29      NGA             209.588        18.1          2.751   
2020-03-01      NGA             209.588        18.1          2.751   
2020-03-02      NGA             209.588        18.1          2.751   
2020-03-03      NGA             209.588        18.1          2.751   
...             ...                 ...         ...            ...   
2022-02-02      NGA             209.588        18.1          2.751   
2022-02-03      NGA             209.588        18.1          2.751   
2022-02-04      NGA             209.588        18.1          2.751   
2022-02-05      NGA             209.588        18.1          2.751   
2022-02-06      NGA             209.588        18.1          2.751   

            aged_70_older  gdp_per_capita  cardiovasc_death_rate  \
date                                                               
2020-02-28          1.447        5338.454                181.013   
2020-02-29          1.447        5338.454                181.013   
2020-03-01          1.447        5338.454                181.013   
2020-03-02          1.447        5338.454                181.013   
2020-03-03          1.447        5338.454                181.013   
...                   ...             ...                    ...   
2022-02-02          1.447        5338.454                181.013   
2022-02-03          1.447        5338.454                181.013   
2022-02-04          1.447        5338.454                181.013   
2022-02-05          1.447        5338.454                181.013   
2022-02-06          1.447        5338.454                181.013   

            handwashing_facilities  life_expectancy  human_development_index  \
date                                                                           
2020-02-28                  41.949            54.69                    0.539   
2020-02-29                  41.949            54.69                    0.539   
2020-03-01                  41.949            54.69                    0.539   
2020-03-02                  41.949            54.69                    0.539   
2020-03-03                  41.949            54.69                    0.539   
...                            ...              ...                      ...   
2022-02-02                  41.949            54.69                    0.539   
2022-02-03                  41.949            54.69                    0.539   
2022-02-04                  41.949            54.69                    0.539   
2022-02-05                  41.949            54.69                    0.539   
2022-02-06                  41.949            54.69                    0.539   

            ...  new_deaths_smoothed_per_million      tests_units  \
date        ...                                                     
2020-02-28  ...                            0.000                0   
2020-02-29  ...                            0.000                0   
2020-03-01  ...                            0.000                0   
2020-03-02  ...                            0.000                0   
2020-03-03  ...                            0.000                0   
...         ...                              ...              ...   
2022-02-02  ...                            0.001  tests performed   
2022-02-03  ...                            0.001  tests performed   
2022-02-04  ...                            0.001  tests performed   
2022-02-05  ...                            0.003  tests performed   
2022-02-06  ...                            0.003  tests performed   

            tests_per_case  positive_rate  new_tests_smoothed  \
date                                                            
2020-02-28             0.0         0.0000                 0.0   
2020-02-29             0.0         0.0000                 0.0   
2020-03-01      

In [138]:
Australia_train

iso_code  population_density  median_age  aged_65_older  \
date                                                                 
2020-01-26      AUS               3.202        37.9         15.504   
2020-01-27      AUS               3.202        37.9         15.504   
2020-01-28      AUS               3.202        37.9         15.504   
2020-01-29      AUS               3.202        37.9         15.504   
2020-01-30      AUS               3.202        37.9         15.504   
...             ...                 ...         ...            ...   
2021-12-31      AUS               3.202        37.9         15.504   
2022-01-01      AUS               3.202        37.9         15.504   
2022-01-02      AUS               3.202        37.9         15.504   
2022-01-03      AUS               3.202        37.9         15.504   
2022-01-04      AUS               3.202        37.9         15.504   

            aged_70_older  gdp_per_capita  extreme_poverty  \
date                                                         
2020-01-26         10.129        44648.71              0.5   
2020-01-27         10.129        44648.71              0.5   
2020-01-28         10.129        44648.71              0.5   
2020-01-29         10.129        44648.71              0.5   
2020-01-30         10.129        44648.71              0.5   
...                   ...             ...              ...   
2021-12-31         10.129        44648.71              0.5   
2022-01-01         10.129        44648.71              0.5   
2022-01-02         10.129        44648.71              0.5   
2022-01-03         10.129        44648.71              0.5   
2022-01-04         10.129        44648.71              0.5   

            cardiovasc_death_rate  hospital_beds_per_thousand  \
date                                                            
2020-01-26                107.791                        3.84   
2020-01-27                107.791                        3.84   
2020-01-28                107.791                        3.84   
2020-01-29                107.791                        3.84   
2020-01-30                107.791                        3.84   
...                           ...                         ...   
2021-12-31                107.791                        3.84   
2022-01-01                107.791                        3.84   
2022-01-02                107.791                        3.84   
2022-01-03                107.791                        3.84   
2022-01-04                107.791                        3.84   

            life_expectancy  ...  new_vaccinations_smoothed_per_million  \
date                         ...                                          
2020-01-26            83.44  ...                                    0.0   
2020-01-27            83.44  ...                                    0.0   
2020-01-28            83.44  ...                                    0.0   
2020-01-29            83.44  ...                                    0.0   
2020-01-30            83.44  ...                                    0.0   
...                     ...  ...                                    ...   
2021-12-31            83.44  ...                                 2883.0   
2022-01-01            83.44  ...                                 2883.0   
2022-01-02            83.44  ...                                 3126.0   
2022-01-03            83.44  ...                                 3200.0   
2022-01-04            83.44  ...                                 4169.0   

            new_people_vaccinated_smoothed  \
date                                         
2020-01-26                             0.0   
2020-01-27                             0.0   
2020-01-28                             0.0   
2020-01-29                             0.0   
2020-01-30                             0.0   
...                                    ...   
2021-12-31                          6752.0   
2022-01-01                          6752.0   
2022-01-02 

### We have tested to see if there are any missing data entries left and there aren't. Each country now can undergo further Feature Engineering.